core/embed/projects/bootloader/protob/protob.c

```diff
@@ -0,0 +1,245 @@
+  MSG_SEND_INIT(Failure);
+  MSG_SEND_ASSIGN_VALUE(code, FailureType_Failure_ActionCancelled);
+  MSG_SEND_ASSIGN_STRING(message, msg);
+  return MSG_SEND(Failure);
+}
+
+secbool send_msg_failure(protob_io_t *iface, FailureType type,
+                         const char *msg) {
+  MSG_SEND_INIT(Failure);
+  MSG_SEND_ASSIGN_VALUE(code, type);
+  MSG_SEND_ASSIGN_STRING(message, msg);
+  return MSG_SEND(Failure);
+}
+
+secbool send_msg_success(protob_io_t *iface, const char *msg) {
+  MSG_SEND_INIT(Success);
+  if (msg != NULL) {
+    MSG_SEND_ASSIGN_STRING(message, msg);
+  }
+  return MSG_SEND(Success);
+}
+
+secbool send_msg_features(protob_io_t *iface, const vendor_header *const vhdr,
+                          const image_header *const hdr) {
+  MSG_SEND_INIT(Features);
+  MSG_SEND_ASSIGN_STRING(vendor, "trezor.io");
+  MSG_SEND_ASSIGN_REQUIRED_VALUE(major_version, VERSION_MAJOR);
+  MSG_SEND_ASSIGN_REQUIRED_VALUE(minor_version, VERSION_MINOR);
+  MSG_SEND_ASSIGN_REQUIRED_VALUE(patch_version, VERSION_PATCH);
+  MSG_SEND_ASSIGN_VALUE(bootloader_mode, true);
+  MSG_SEND_ASSIGN_STRING(model, MODEL_NAME);
+  MSG_SEND_ASSIGN_STRING(internal_model, MODEL_INTERNAL_NAME);
+  if (vhdr && hdr) {
+    MSG_SEND_ASSIGN_VALUE(firmware_present, true);
+    MSG_SEND_ASSIGN_VALUE(fw_major, (hdr->version & 0xFF));
+    MSG_SEND_ASSIGN_VALUE(fw_minor, ((hdr->version >> 8) & 0xFF));
+    MSG_SEND_ASSIGN_VALUE(fw_patch, ((hdr->version >> 16) & 0xFF));
+    MSG_SEND_ASSIGN_STRING_LEN(fw_vendor, vhdr->vstr, vhdr->vstr_len);
+  } else {
+    MSG_SEND_ASSIGN_VALUE(firmware_present, false);
+  }
+  if (unit_properties()->color_is_valid) {
+    MSG_SEND_ASSIGN_VALUE(unit_color, unit_properties()->color);
+  }
+  if (unit_properties()->packaging_is_valid) {
+    MSG_SEND_ASSIGN_VALUE(unit_packaging, unit_properties()->packaging);
+  }
+  if (unit_properties()->btconly_is_valid) {
+    MSG_SEND_ASSIGN_VALUE(unit_btconly, unit_properties()->btconly);
+  }
+
+#if USE_OPTIGA
+  MSG_SEND_ASSIGN_VALUE(bootloader_locked,
+                        (secret_bootloader_locked() == sectrue));
+#endif
+  return MSG_SEND(Features);
+}
+
+secbool recv_msg_initialize(protob_io_t *iface, Initialize *msg) {
+  MSG_RECV_INIT(Initialize);
+  secbool result = MSG_RECV(Initialize);
+  memcpy(msg, &msg_recv, sizeof(Initialize));
+  return result;
+}
+
+secbool recv_msg_get_features(protob_io_t *iface, GetFeatures *msg) {
+  MSG_RECV_INIT(GetFeatures);
+  secbool result = MSG_RECV(GetFeatures);
+  memcpy(msg, &msg_recv, sizeof(GetFeatures));
+  return result;
+}
+
+secbool recv_msg_ping(protob_io_t *iface, Ping *msg) {
+  MSG_RECV_INIT(Ping);
+  secbool result = MSG_RECV(Ping);
+  memcpy(msg, &msg_recv, sizeof(Ping));
+  return result;
+}
+
+secbool recv_msg_firmware_erase(protob_io_t *iface, FirmwareErase *msg) {
+  MSG_RECV_INIT(FirmwareErase);
+  secbool result = MSG_RECV(FirmwareErase);
+  memcpy(msg, &msg_recv, sizeof(FirmwareErase));
+  return result;
+}
```

Comment:
there's a bunch of inconsistencies: we have `recv_msg_get_features` (empty message) but not `recv_msg_wipe_device` (also empty message)

furthermore these are cookie cutter functions, that is to say, macroable.

i would expect the main `wf_host_control` function would always `recv_msg_xyz` and hand the result over to the respective workflow?

Reply:
Added the `recv_msg_wipe_device` for consistency in https://github.com/trezor/trezor-firmware/commit/c421c81c9e1465fd5716ffabeafb666ea27a0e29

I don't have a way now to universally receive any message in `wf_host_control`, there could be also some challenges with larger messages that require streaming. I agree that this would be a better way, but i consider this somewhat out of originally intended scope of refactoring, and as it is already large enough i'd leave it for future PRs.